src/main/groovy/org/pih/warehouse/inventory/CycleCountRequestCommand.groovy

```diff
@@ -0,0 +1,33 @@
+package org.pih.warehouse.inventory
+
+import grails.validation.Validateable
+import org.pih.warehouse.core.Location
+import org.pih.warehouse.product.Product
+
+class CycleCountRequestCommand implements Validateable {
+
+    Location facility
+
+    List<Product> products
+
+    Boolean blindCount
+
+    static constraints = {
+        products(nullable: true, validator: { List<Product> products, CycleCountRequestCommand obj ->
+            String productCode = null
+            boolean duplicateExists = products.any { Product product ->
+                // FIXME: Most probably, this will have to be modified not to include completed cycle counts
+                CycleCountRequest cycleCountRequest = CycleCountRequest.findByProductAndFacility(product, obj.facility)
+                if (cycleCountRequest) {
+                    productCode = product.productCode
```

Comment:
using productCode causes this to do a lookup on product for each product in the list, correct? I was hoping we could do this validation entirely on productId (which I was assuming would avoid the lookup since we already have the ids in the request body).

But actually this is making me realize I don't know how Grails is binding domains to request/command objects. Is it smart enough to not pre-fetch the products until they're used?

Reply:
`Product` is bound at the binding step, so at this point nothing is fetched. It also doesn't cause any N+1, because `productCode` is a string field. Potential N+1 would be caused, if I was accessing a collection here.

Aside: I only store and assign the `productCode` here for a user-friendly validation message, so that the user knows which product it is about (id is not always user-friendly for products, because it's not always id == productCode)

Reply:
And also I didn't see a point to have the property `String productId` in the command instead of `Product product`, because anyway we'd have to fetch the product at some point, and here, Grails does it automatically for us.

Reply:
this also gives us much flexibility while validating in the constraints and in the validator, because we can already operate on the product and its properties, instead of validating the `String` property, where the only possibilities would be to validate if it's null/blank.

Reply:
I was assuming we could bypass fetching the Product at the binding step entirely and just use the id since Product isn't actually used anywhere (except your error case but I'm fine with fetching it there). All we do is map it to the request.

But thinking about it, we do need to verify that the given id is pointing to a real product, so we'll need the lookup regardless.

Okay I'm convinced.